### PR TITLE
[WebApp] Add challenge sudo mode

### PR DESCRIPTION
**Pull request by Maks19**
**Description:** Add challenge sudo mode for Google SSO and PayPal

**Issue:** https://www.notion.so/saladtech/FE-Intercept-failed-requests-for-Paypal-and-Google-SSO-52b362dd2de3430c97fbb51b3baacdef

### CODE REVIEW

File: packages/web-app/src/modules/protected-action/utils.tsx [hunk 1]
```diff
@@ -34,6 +38,19 @@ export const handlePendingProtectedAction = (store: RootStore) => {
...
+        case ChallengeSudoModeTrigger.GoogleSignIn:
+          store.routing.push('/account/summary', { isGoogleSignInFormTriggered: true })
+          break
+        case ChallengeSudoModeTrigger.PayPalLogIn:
+          window.open(config.paypalUrl)
+          store.routing.push('/account/summary', { isPayPalLogInTriggered: true })
+          break
+      }
+      store.auth.setChallengeSudoModeTrigger(undefined)
+      break
```
Comment (vitto-moz) [inline packages/web-app/src/modules/protected-action/utils.tsx:53]: why cannot we make this condition handled plain - so just add `ChallengeSudoModeTrigger` enum values into `ChallengeSudoModeTrigger`

File: packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx [hunk 2]
```diff
@@ -53,34 +51,54 @@ const styles = {
...
+  challengeSudoMode,
 }) => {
   const shouldEnableGoogleSignInButton = !isTermsAndConditionsRequired || isTermsAndConditionsAccepted
   const shouldDisableGoogleSignInButton = isTermsAndConditionsRequired && !isTermsAndConditionsAccepted
 
+  const hiddenFormRef = useRef<HTMLFormElement>(null)
+
+  const handleSignInWithGoogle = useCallback(async () => {
+    try {
+      const response = await challengeSudoMode(ChallengeSudoModeTrigger.GoogleSignIn)
```
Comment (vitto-moz) [inline packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx:71]: we should not execute requests and handle responses in a component - let's use store for that

**Review by vitto-moz — CHANGES_REQUESTED** [review]

File: packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx (hunk 2 above)
Comment (Maks19) [reply packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx]: Fixed

File: packages/web-app/src/modules/protected-action/utils.tsx (hunk 1 above)
Comment (Maks19) [reply packages/web-app/src/modules/protected-action/utils.tsx]: Fixed

File: packages/web-app/src/modules/account-views/account-views/components/PayPalLoginButton.tsx [hunk 3]
```diff
@@ -1,26 +1,35 @@
...
-interface PayPalLoginButtonProps extends WithStyles<typeof styles> {
-  onClick: () => void
-}
+export const PayPalLoginButton: FunctionComponent<PayPalLoginButtonProps> = ({
+  handleCheckPayPalId,
+  challengeSudoMode,
+}) => {
+  const handlePaypalLogInClick = async () => {
+    try {
+      const response = await challengeSudoMode(ChallengeSudoModeTrigger.PayPalLogIn)
```
[the remark this answers is not in the document]
Comment (Maks19) [reply packages/web-app/src/modules/account-views/account-views/components/PayPalLoginButton.tsx]: Fixed

File: packages/web-app/src/modules/account-views/account-views/components/GoogleHiddenSignInForm.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  classes,
+  isTermsAndConditionsAccepted,
+  isTermsAndConditionsRequired,
+  hiddenFormRef,
+}) => (
+  <>
+    <form
+      className={classes.hiddenFormContainer}
+      action={`${config.apiBaseUrl}/api/v2/authentication/external/account`}
+      ref={hiddenFormRef}
```
Comment (vitto-moz) [inline packages/web-app/src/modules/account-views/account-views/components/GoogleHiddenSignInForm.tsx:30]: shouldn't we use `React.forwardRef` to make it work correct?

File: packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx [hunk 5]
```diff
@@ -53,34 +49,49 @@ const styles = {
...
+  googleSignInChallengeSudoMode,
 }) => {
   const shouldEnableGoogleSignInButton = !isTermsAndConditionsRequired || isTermsAndConditionsAccepted
   const shouldDisableGoogleSignInButton = isTermsAndConditionsRequired && !isTermsAndConditionsAccepted
 
+  const hiddenFormRef = useRef<HTMLFormElement>(null)
+
+  const handleHiddenFormSubmit = useCallback(() => {
+    if (hiddenFormRef.current) {
+      googleSignInChallengeSudoMode(hiddenFormRef.current.submit.bind(hiddenFormRef.current))
```
Comment (vitto-moz) [inline packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx:69]:
```suggestion
      const signInWithGoogle = hiddenFormRef.current.submit.bind(hiddenFormRef.current)
      googleSignInChallengeSudoMode(signInWithGoogle )
```

File: packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx [hunk 6]
```diff
@@ -53,34 +49,49 @@ const styles = {
 export interface GoogleSignInFormProps extends WithStyles<typeof styles> {
   isTermsAndConditionsAccepted: boolean
   isTermsAndConditionsRequired: boolean
+  isGoogleSignInFormTriggered: boolean
+  googleSignInChallengeSudoMode: (googleSignIn: () => void) => void
```
Comment (vitto-moz) [inline packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx:53]:
```suggestion
  googleSignInChallengeSudoMode: (signInWithGoogle: () => void) => void
```

File: packages/web-app/src/modules/auth/AuthStore.tsx [hunk 7]
```diff
@@ -69,4 +87,49 @@ export class AuthStore {
...
+      const response = yield this.axios.post(authenticationSessionsSudoEndpointPath)
+      return response
+    } catch (error) {
+      this.challengeSudoModeTrigger = challengeSudoModeTrigger
+      console.error('AuthStore -> challengeSudoMode: ', error)
+      return null
+    }
+  })
+
+  googleSignInChallengeSudoMode = async (googleSignIn: () => void): Promise<void> => {
```
Comment (vitto-moz) [inline packages/web-app/src/modules/auth/AuthStore.tsx:103]:
```suggestion
  signInWithGoogleChallengeSudoMode= async (signInWithGoogle: () => void): Promise<void> => {
```

File: packages/web-app/src/modules/auth/AuthStore.tsx [hunk 8]
```diff
@@ -69,4 +87,49 @@ export class AuthStore {
...
+      const response = await this.challengeSudoMode(ChallengeSudoModeTrigger.GoogleSignIn)
+      if (response) {
+        googleSignIn()
+      }
+    } catch (error) {
+      console.error('AuthStore -> googleSignInChallengeSudoMode: ', error)
+    }
+  }
+
+  payPalLogInChallengeSudoMode = async (): Promise<void> => {
```
Comment (vitto-moz) [inline packages/web-app/src/modules/auth/AuthStore.tsx:114]:
```suggestion
  sighInWithPayPalChallengeSudoMode= async (): Promise<void> => {
```

File: packages/web-app/src/modules/profile/ProfileStore.tsx [hunk 9]
```diff
@@ -23,11 +23,7 @@ const SaladDefaultAvatar: Avatar = {
 const installReminderFeatureReleaseDate = new Date('2023-12-21T17:10:47.324Z')
 const IS_INSTALL_REMINDER_CLOSED_STORAGE_KEY = 'IS_INSTALL_REMINDER_CLOSED'
 
-interface PendingProtectedAction {
-  method: string
-  url: string
-  data?: string
-}
+let timeoutId: NodeJS.Timeout
```
Comment (vitto-moz) [inline packages/web-app/src/modules/profile/ProfileStore.tsx:26]: let's make it a private property named `loadPayPalIdTimeout` of a class `ProfileStore` and place near `loadPayPalId` property

**Review by vitto-moz — CHANGES_REQUESTED** [review]

File: packages/web-app/src/modules/profile/ProfileStore.tsx (hunk 9 above)
Comment (Maks19) [reply packages/web-app/src/modules/profile/ProfileStore.tsx:26]: Fixed

File: packages/web-app/src/modules/auth/AuthStore.tsx (hunk 8 above)
Comment (Maks19) [reply packages/web-app/src/modules/auth/AuthStore.tsx:114]: Fixed

File: packages/web-app/src/modules/auth/AuthStore.tsx (hunk 7 above)
Comment (Maks19) [reply packages/web-app/src/modules/auth/AuthStore.tsx:103]: Fixed

File: packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx (hunk 6 above)
Comment (Maks19) [reply packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx:53]: Fixed

File: packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx (hunk 5 above)
Comment (Maks19) [reply packages/web-app/src/modules/account-views/account-views/components/GoogleSignInForm.tsx:69]: Fixed

File: packages/web-app/src/modules/account-views/account-views/components/GoogleHiddenSignInForm.tsx (hunk 4 above)
Comment (Maks19) [reply packages/web-app/src/modules/account-views/account-views/components/GoogleHiddenSignInForm.tsx:30]: Fixed

File: packages/web-app/src/modules/account-views/account-views/AccountContainer.tsx [hunk 10]
```diff
@@ -29,6 +29,9 @@ const mapStoreToProps = (store: RootStore): any => ({
   isTermsAndConditionsAccepted: store.termsAndConditions.isTermsAndConditionsAccepted,
   onSubmitTermsAndConditions: store.termsAndConditions.submitTermsAndConditions,
   onToggleAcceptTermsAndConditions: store.termsAndConditions.toggleAcceptTermsAndConditions,
+  checkPayPalIdWithInterval: store.profile.checkPayPalIdWithInterval,
+  googleSignInChallengeSudoMode: store.auth.googleSignInChallengeSudoMode,
+  payPalLogInChallengeSudoMode: store.auth.payPalLogInChallengeSudoMode,
```
[the remark this answers is not in the document]
Comment (Maks19) [reply packages/web-app/src/modules/account-views/account-views/AccountContainer.tsx:34]: Fixed
used `logInWithPayPalChallengeSudoMode`

File: packages/web-app/src/modules/protected-action/ProtectedActionPageContainer.tsx [hunk 11]
```diff
@@ -4,7 +4,7 @@ import { ProtectedActionPage } from './components'
 
 const mapStoreToProps = (store: RootStore): any => ({
   isPasskeySupported: store.passkey.isPasskeySupported,
-  isPendingProtectedActionExist: !!store.profile.pendingProtectedAction,
+  isPendingProtectedActionExist: !!store.auth.pendingProtectedAction,
```
[the remark this answers is not in the document]
Comment (Maks19) [reply packages/web-app/src/modules/protected-action/ProtectedActionPageContainer.tsx:7]: Changed it

**Review by vitto-moz — APPROVED** [review]

**Pull request merged**